File: 62. Unique Paths/Solution.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # m = number of rows
        # n = number of columns

        grid = [[0 for x in range(n)] for x in range(m)] # dual list comprehension
        """
        for a 3x4 (m x n):
        [[0, 0, 0, 0],
        [0, 0, 0, 0],
        [0, 0, 0, 0]]
        """

        # return the number of unique paths in the right cell and below cell
        def checkPaths(i, j):
            # base case
            # if at the start/top left corner, return 1
            if j == 0 and i == 0:
                return 1
            
            accum = 0 # number of unique paths to this cell
            # out-of-bounds checking
            if i < m: # columns (going too far down)
                accum += grid[i-1][j]
            if j < n: # rows (going too far right)
                accum += grid[i][j-1]
            return accum
        
        for i in range(m): # rows (going down)
            for j in range(n): # columns (going right)
                grid[i][j] = checkPaths(i, j)

        print(grid)
        return grid[-1][-1] # print the finish cell value

File: 62. Unique Paths/Solution.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        // m = number of rows
        // n = number of columns
        vector<vector<int>> grid(m, vector<int>(n, 0));
        return calcPaths(0, 0, m, n, grid);
    }
    int calcPaths(int i, int j, int m, int n, vector<vector<int>>& grid){
        if (i == m || j == n){ // out of bounds
            return 0;
        } 
        if (i == m-1 && j == n-1){ // bottom right corner
            return 1; 
        }
        if (grid[i][j] == 0){ // not calculated yet
            grid[i][j] = calcPaths(i+1, j, m, n, grid) + calcPaths(i, j+1, m, n, grid);
        }
        return grid[i][j];  
    }
};